src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -218,6 +256,81 @@ public UnfilteredRowIterator executePartitionReadCommand(SinglePartitionReadComm
+    {
+        var filterOperation = filterOperation();
+        var orderings = filterOperation.expressions()
+                                       .stream().filter(e -> e.operator() == Operator.ANN).collect(Collectors.toList());
+        assert orderings.size() <= 1;
+        if (filterOperation.expressions().size() == 1 && filterOperation.children().isEmpty() && orderings.size() == 1)
+            // If we only have one expression, we just use the ANN index to order and limit.
+            return getTopKRows(orderings.get(0));
+
+        // We already decided we need to do first sort then filter, so no need to open the index iterator:
+        if (queryContext.filterSortOrder() == QueryContext.FilterSortOrder.SORT_THEN_FILTER)
+        {
+            assert queryContext.postFilterSelectivityEstimate() != null;
+            return getTopKRows(orderings.get(0));
+        }
+
+        var nonOrderingExpressions = filterOperation.expressions().stream()
+                                                    .filter(e -> e.operator() != Operator.ANN)
+                                                    .collect(Collectors.toList());
+        var iter = Operation.Node.buildTree(nonOrderingExpressions, filterOperation.children(), filterOperation.isDisjunction()).analyzeTree(this).rangeIterator(this);
+
+        if (orderings.isEmpty())
+            return iter;
+
+        if (queryContext.filterSortOrder() == null)
+        {
+            QueryContext.FilterSortOrder order = decideFilterSortOrder(nonOrderingExpressions, iter);
+            queryContext.setFilterSortOrder(order);
+        }
+        
+        if (queryContext.filterSortOrder() == QueryContext.FilterSortOrder.SORT_THEN_FILTER)
+        {
+            queryContext.setPostFilterSelectivityEstimate(estimateSelectivity(iter));
+            FileUtils.closeQuietly(iter);
+            return getTopKRows(orderings.get(0));
+        }
+
+        return getTopKRows(iter, orderings.get(0));
+    }
+
+    private QueryContext.FilterSortOrder decideFilterSortOrder(List<RowFilter.Expression> filterExpressions, RangeIterator iter)
+    {
+        double sortThenFilterCost = estimateSortThenFilterCost(iter);
+        double filterThenSortCost = estimateFilterThenSortCost(filterExpressions, iter);
+        QueryContext.FilterSortOrder order = sortThenFilterCost < filterThenSortCost
+               ? QueryContext.FilterSortOrder.SORT_THEN_FILTER
+               : QueryContext.FilterSortOrder.FILTER_THEN_SORT;
+        logger.debug("Decided filter sort order {} (costs: SORT_THEN_FILTER = {}, FILTER_THEN_SORT = {})",
+                     order, sortThenFilterCost, filterThenSortCost);
+        return order;
+    }
+
+    private double estimateFilterThenSortCost(List<RowFilter.Expression> filterExpressions, RangeIterator iter)
+    {
+        long primaryKeysFetchedFromIndex = iter.getMaxKeys();
+        float intersectionPenalty = filterExpressions.size() > 1 ? Costs.INDEX_INTERSECTION_PENALTY : 1.0f;
```

Comment:
I think we want slightly more robust logic here using the `filterOperation.children()` to properly catch all intersections. This logic only catches top level intersections. For example, the following won't get classified as an intersection `WHERE x < 5 OR (y = 5 AND x =2) ORDER BY...` because the top level of the `filterOperation` only has the ANN expression.

On that note, should we have a "penalty" for the number of predicates to indicate the extra cost to compute the result? Even if it is naive, it seems reasonable to assume the more operations we have, the more costly the filter then sort will be.

Reply:
fixed; unions are cheap so I don't count them for now, but now nested intersections are counted